### ISSUE TITLE
Generator send types not applied to yield expressions

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#1652]
### Describe the bug

When implementing a generator with a send type, the value of a `yield` expression is of an unknown type.

### To reproduce

Taking the [generator typing example from the Python docs](https://docs.python.org/3/library/typing.html#typing.Generator), the type of `sent` is unknown:

```python
from typing import Generator

def echo_round() -> Generator[int, float, str]:
    sent = yield 0
    while sent >= 0:
        sent = yield round(sent)
    return 'Done'
```

### Expected behavior

`sent` should be considered a `float`, matching the generator type hint.

NB. `yield from` expressions do correctly gain the generator's return type, i.e. the expression `yield from echo_round()` is given the `str` type as expected (though of course this example isn't designed to be yielded from in that way).

### VS Code extension or command-line

Pyright 1.1.122 (or via Pylance 2021.3.2) in VSCode.

**Comment by erictraut** [comment]
Thanks for the bug report. Your analysis is correct. I was using the same logic for "yield" and "yield from", but it needs to be different to produce the correct results. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.123, which I just published. It will also be included in the next release of pylance.

**Comment by Terrance** [comment]
It seems like there is a case where a `yield from` expression doesn't get a type either:

```python
class Yielder:
    def __iter__(self) -> Generator[int, None, bool]:
        yield 1
        return True

def collect():
    y = Yielder()
    ok = yield from y  # unknown type
    return ok
```

Providing the generator directly with `yield from y.__iter__()` gains the `bool` type as expected.

At a guess, it's being cast to `Iterator[int]` somewhere as part of the `__iter__` protocol.  Is it possible to have the return type available here, or is this more of a Python limitation?
